### ISSUE TITLE
Resulting sparsity is 0 for all the layers

### ISSUE THREAD

**Issue by Tramtadama** [neuralmagic/sparseml#1675]
Hi,
I have tried to prune two different models, one for image regression (resnet) and another for segmentation (unet, resnet backbone).
Both after finished with pruning return 0 for all layers with tensor_sparsity(). Tried running the beans tutorial and there the pruning worked.
Here is my recipe for the segmentation:
```
# Epoch hyperparams
stabilization_epochs: 1.0
pruning_epochs: 9.0
finetuning_epochs: 5.0

# Learning rate hyperparams
init_lr: 0.0001
final_lr: 0.00005

# Pruning hyperparams
init_sparsity: 0.05
final_sparsity: 0.9

# Stabalization Stage
training_modifiers:
  - !EpochRangeModifier
    start_epoch: 0.0
    end_epoch: eval(stabilization_epochs + pruning_epochs + finetuning_epochs)

  - !SetLearningRateModifier
    start_epoch: 0.0
    learning_rate: eval(init_lr)

# Pruning Stage
pruning_modifiers:
  - !LearningRateFunctionModifier
    init_lr: eval(init_lr)
    final_lr: eval(final_lr)
    lr_func: cosine
    start_epoch: eval(stabilization_epochs)
    end_epoch: eval(stabilization_epochs + pruning_epochs)

  - !GlobalMagnitudePruningModifier
    init_sparsity: eval(init_sparsity)
    final_sparsity: eval(final_sparsity)
    start_epoch: eval(stabilization_epochs)
    end_epoch: eval(stabilization_epochs + pruning_epochs)
    update_frequency: 0.5
    params:
        - 'encoder._conv_stem.weight'
        - 'encoder._blocks.0._depthwise_conv.weight'
        - 'encoder._blocks.0._se_reduce.weight'
        - 'encoder._blocks.0._se_expand.weight'
        - 'encoder._blocks.0._project_conv.weight'
        - 'encoder._blocks.1._expand_conv.weight'
        - 'encoder._blocks.1._depthwise_conv.weight'
        - 'encoder._blocks.1._se_reduce.weight'
        - 'encoder._blocks.1._se_expand.weight'
        - 'encoder._blocks.1._project_conv.weight'
        - 'encoder._blocks.2._expand_conv.weight'
        - 'encoder._blocks.2._depthwise_conv.weight'
        - 'encoder._blocks.2._se_reduce.weight'
        - 'encoder._blocks.2._se_expand.weight'
        - 'encoder._blocks.2._project_conv.weight'
        - 'encoder._blocks.3._expand_conv.weight'
        - 'encoder._blocks.3._depthwise_conv.weight'
        - 'encoder._blocks.3._se_reduce.weight'
        - 'encoder._blocks.3._se_expand.weight'
        - 'encoder._blocks.3._project_conv.weight'
        - 'encoder._blocks.4._expand_conv.weight'
        - 'encoder._blocks.4._depthwise_conv.weight'
        - 'encoder._blocks.4._se_reduce.weight'
        - 'encoder._blocks.4._se_expand.weight'
        - 'encoder._blocks.4._project_conv.weight'
        - 'encoder._blocks.5._expand_conv.weight'
        - 'encoder._blocks.5._depthwise_conv.weight'
        - 'encoder._blocks.5._se_reduce.weight'
        - 'encoder._blocks.5._se_expand.weight'
        - 'encoder._blocks.5._project_conv.weight'
        - 'encoder._blocks.6._expand_conv.weight'
        - 'encoder._blocks.6._depthwise_conv.weight'
        - 'encoder._blocks.6._se_reduce.weight'
        - 'encoder._blocks.6._se_expand.weight'
        - 'encoder._blocks.6._project_conv.weight'
        - 'encoder._blocks.7._expand_conv.weight'
        - 'encoder._blocks.7._depthwise_conv.weight'
        - 'encoder._blocks.7._se_reduce.weight'
        - 'encoder._blocks.7._se_expand.weight'
        - 'encoder._blocks.7._project_conv.weight'
        - 'encoder._blocks.8._expand_conv.weight'
        - 'encoder._blocks.8._depthwise_conv.weight'
        - 'encoder._blocks.8._se_reduce.weight'
        - 'encoder._blocks.8._se_expand.weight'
        - 'encoder._blocks.8._project_conv.weight'
        - 'encoder._blocks.9._expand_conv.weight'
        - 'encoder._blocks.9._depthwise_conv.weight'
        - 'encoder._blocks.9._se_reduce.weight'
        - 'encoder._blocks.9._se_expand.weight'
        - 'encoder._blocks.9._project_conv.weight'
        - 'encoder._blocks.10._expand_conv.weight'
        - 'encoder._blocks.10._depthwise_conv.weight'
        - 'encoder._blocks.10._se_reduce.weight'
        - 'encoder._blocks.10._se_expand.weight'
        - 'encoder._blocks.10._project_conv.weight'
        - 'encoder._blocks.11._expand_conv.weight'
        - 'encoder._blocks.11._depthwise_conv.weight'
        - 'encoder._blocks.11._se_reduce.weight'
        - 'encoder._blocks.11._se_expand.weight'
        - 'encoder._blocks.11._project_conv.weight'
        - 'encoder._blocks.12._expand_conv.weight'
        - 'encoder._blocks.12._depthwise_conv.weight'
        - 'encoder._blocks.12._se_reduce.weight'
        - 'encoder._blocks.12._se_expand.weight'
        - 'encoder._blocks.12._project_conv.weight'
        - 'encoder._blocks.13._expand_conv.weight'
        - 'encoder._blocks.13._depthwise_conv.weight'
        - 'encoder._blocks.13._se_reduce.weight'
        - 'encoder._blocks.13._se_expand.weight'
        - 'encoder._blocks.13._project_conv.weight'
        - 'encoder._blocks.14._expand_conv.weight'
        - 'encoder._blocks.14._depthwise_conv.weight'
        - 'encoder._blocks.14._se_reduce.weight'
        - 'encoder._blocks.14._se_expand.weight'
        - 'encoder._blocks.14._project_conv.weight'
        - 'encoder._blocks.15._expand_conv.weight'
        - 'encoder._blocks.15._depthwise_conv.weight'
        - 'encoder._blocks.15._se_reduce.weight'
        - 'encoder._blocks.15._se_expand.weight'
        - 'encoder._blocks.15._project_conv.weight'
        - 'encoder._conv_head.weight'
        - 'decoder.blocks.0.conv1.0.weight'
        - 'decoder.blocks.0.conv2.0.weight'
        - 'decoder.blocks.1.conv1.0.weight'
        - 'decoder.blocks.1.conv2.0.weight'
        - 'decoder.blocks.2.conv1.0.weight'
        - 'decoder.blocks.2.conv2.0.weight'
        - 'decoder.blocks.3.conv1.0.weight'
        - 'decoder.blocks.3.conv2.0.weight'
        - 'decoder.blocks.4.conv1.0.weight'
        - 'decoder.blocks.4.conv2.0.weight'
        - 'segmentation_head.0.weight'
    leave_enabled: True

# Finetuning Stage
finetuning_modifiers:
  - !LearningRateFunctionModifier
    init_lr: eval(init_lr)
    final_lr: eval(final_lr)
    lr_func: cosine
    start_epoch: eval(stabilization_epochs + pruning_epochs)
    end_epoch: eval(stabilization_epochs + pruning_epochs + finetuning_epochs) 
```
Only things that I add to my regular code are:
```
manager = ScheduledModifierManager.from_yaml('sparsify_recipe.yaml')
optimizer = manager.modify(model, optim, steps_per_epoch=len(train_loader))
```
and at the end,
```
manager.finalize(model)
```
After that I save model with ```torch.save()``` as tar file. After loading and running ```tensor_sparsity()``` it returns 0.000 for all the layers.

Not sure if this is a bug necessarily, maybe I do something wrong that I am not aware of.

Thank you.

**Comment by jeanniefinks** [comment]
What is your model definition and what the intergration to the training code looks like?

**Comment by Tramtadama** [comment]
Thank you for your reply, here is my code:
```
train_batch_size = 64
val_batch_size = 1

shape = (224, 224)

train_loader = DataLoader(
    train_dset,
    batch_size=train_batch_size,
    shuffle=True,
    num_workers=6)

lr = 0.0005

model = smp.Unet('efficientnet-b0', activation=None, encoder_weights=None)
model = model.cuda()

optim = torch.optim.Adam(model.parameters(), lr=lr)

epochs = 50
no_batches_val = math.ceil(val_size / val_batch_size)
loss = BCEDiceLoss()
no_batches_train = math.ceil(train_size / train_batch_size)

checkpoint = torch.load('efbn0_224_sparsify_test(14).tar')
model.load_state_dict(checkpoint['model_state_dict'])

manager = ScheduledModifierManager.from_yaml('sparsify_recipe.yaml')
optimizer = manager.modify(model, optim, steps_per_epoch=len(train_loader))
train_loss_history = []
train_iou_history = []
for epoch in tqdm(range(manager.max_epochs)):

    model = model.cuda().train()

    for (imgs, labels, _) in tqdm(train_loader):
        optim.zero_grad()

        imgs, labels = imgs.cuda(), labels.cuda()

        out = model(imgs)
        batch_train_loss = loss(out, labels)

        batch_train_loss.backward()
        optim.step()
    
manager.finalize(model)

torch.save({
            'epoch': epoch,
            'model_state_dict': model.state_dict(),
            },  'efbn0_224_sparse_final(' + str(epoch) + ').tar')

```

**Comment by Satrat** [comment]
Hi @Tramtadama, I'll take a look into this and see if I can reproduce. A few questions:
* Could you clarify what dataset you are using? I see you're passing `train_dset` to a DataLoader but your code doesn't include the definition
* Could you also include the definition for `BCEDiceLoss()`?
* Is this example supposed to be for image segmentation or classification? The model you're loading seems to be for segmentation, but you're calculating loss on labels

**Comment by Tramtadama** [comment]
Hi @Satrat, thank you for you efforts. The task is binary segmentation. The "label" is 224x224 image mask of 0's and 1's.
Here is the definition of the loss function:
```
import torch
from torch import nn

def f_score(pr, gt, beta=1, eps=1e-7, threshold=None, activation='sigmoid'):
    """
    Args:
        pr (torch.Tensor): A list of predicted elements
        gt (torch.Tensor):  A list of elements that are to be predicted
        beta (float): positive constant
        eps (float): epsilon to avoid zero division
        threshold: threshold for outputs binarization
    Returns:
        float: F score
    """

    if activation is None or activation == "none":
        activation_fn = lambda x: x
    elif activation == "sigmoid":
        activation_fn = torch.nn.Sigmoid()
    elif activation == "softmax2d":
        activation_fn = torch.nn.Softmax2d()
    else:
        raise NotImplementedError(
            "Activation implemented for sigmoid and softmax2d"
        )

    pr = activation_fn(pr)

    if threshold is not None:
        pr = (pr > threshold).float()


    tp = torch.sum(gt * pr)
    fp = torch.sum(pr) - tp
    fn = torch.sum(gt) - tp

    score = ((1 + beta ** 2) * tp + eps) \
            / ((1 + beta ** 2) * tp + beta ** 2 * fn + fp + eps)

class DiceLoss(nn.Module):
    __name__ = 'dice_loss'

    def __init__(self, eps=1e-7, activation='sigmoid'):
        super().__init__()
        self.activation = activation
        self.eps = eps

    def forward(self, y_pr, y_gt):
        return 1 - f_score(y_pr, y_gt, beta=1., eps=self.eps, threshold=None, activation=self.activation)```

class BCEDiceLoss(DiceLoss):
    __name__ = 'bce_dice_loss'

    def __init__(self, eps=1e-7, activation='none'):
        super().__init__(eps, activation)
        self.bce = nn.BCEWithLogitsLoss(reduction='mean')

    def forward(self, y_pr, y_gt):
        dice = super().forward(y_pr, y_gt)
        bce = self.bce(y_pr, y_gt)
        return dice + bce
```
Hope this helps with the reproduction, if there is anything more that you need please tell me. The dataset is private so I do not know if providing the definition helps? It is an image of a car with license plate and corresponding binary mask where the area of license plate are 1's and everything else is 0, there are also images where there is no license plate so corresponding mask is just 0's.
Thank you!

**Comment by Tramtadama** [comment]
Fixed the optimizer mistake, now its working. Sorry for the inconvenience.
